### PR TITLE
Documentation (index.html) and tests for client-side alpn

### PR DESCRIPTION
**Pull request by zellerin**
Documentation and tests for #149.

The tests were tested locally only, not across full spectrum of Lisps and SSL versions.

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -292,6 +292,11 @@ <h3>API</h3>
       to choose certificate for right domain. Also the <tt>hostname></tt> is used for
       hostname verification if verification is enabled by <tt>verify</tt>.
     </p>
+    <p>
+      <tt>alpn-protocols</tt> if specified, should be a list of alpn protocol
+      names that would be offered to the server. Protocol selected by the server
+      can be retrieved by <tt>get-selected-alpn-protocol</tt>.
+    </p>
```
Comment (avodonosov) [inline index.html:299]: It would be good also to add description of the `cl+ssl:get-selected-alpn-protocol`

File: cl+ssl.test.asd [hunk 2]
```diff
@@ -27,4 +27,5 @@
                  (:file "fingerprint")
                  (:file "verify-hostname")
                  (:file "badssl-com")
-                 (:file "bio")))))
+                 (:file "bio")
+		 (:file "alpn")))))
```
Comment (avodonosov) [inline cl+ssl.test.asd:31]: Could you please indent as other files above

File: test/alpn.lisp [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+(in-suite :cl+ssl.alpn)
+
+(test alpn-client
+  (when (cl+ssl::openssl-is-at-least 1 0 2)
+    (flet ((test-alpn-result (target proposed expected)
+	     (usocket:with-client-socket (socket stream target 443)
+	       (is
+		(equal expected
+		       (cl+ssl:get-selected-alpn-protocol
+			(cl+ssl:make-ssl-client-stream stream :alpn-protocols proposed)))))))
```
Comment (avodonosov) [inline test/alpn.lisp:15]: strange indentation (at least in the github web UI)

**Review by avodonosov — CHANGES_REQUESTED** [review]

File: test/alpn.lisp (hunk 3 above)
Comment (zellerin) [reply test/alpn.lisp:15]: Tabs problems, in both cases. Will fix.

File: index.html (hunk 1 above)
Comment (zellerin) [reply index.html:299]: Ok.

**Comment by avodonosov** [comment]
Merged. thank you.

**Comment by avodonosov** [comment]
@zellerin, somehow the alpn test fails on ABCL, http://cl-test-grid.appspot.com/blob?key=6xx13j9zqd

**Comment by zellerin** [comment]
@avodonosov 
looks like an issue in test itself, the socket must be binary for abcl to work:
```
             (usocket:with-client-socket (socket stream target 443
                                                 :element-type '(unsigned-byte 8))
```
The test passes after the change when run locally. Will prepare patch later this weekend if needed/confirmed.

**Comment by avodonosov** [comment]
@zellerin , one more question.

When I run the testsuite loally, I have the alpn test failed like this:
```
 Did 67 checks.
    Pass: 66 (98%)
    Skip: 0 ( 0%)
    Fail: 1 ( 1%)

 Failure Details:
 --------------------------------
 ALPN-CLIENT in CL+SSL.ALPN []: 
      Unexpected Error: #<CL+SSL:SSL-ERROR-VERIFY #x302003DA4ECD>
SSL verify error: 20 X509_V_ERR_UNABLE_TO_GET_ISSUER_CERT_LOCALLY.
 --------------------------------
```

Do you have this now?

**Comment by avodonosov** [comment]
My failure was on CCL

**Comment by avodonosov** [comment]
Turned out my failure was only happening when I use OpenSSL binaries I build myself for CI and does not happen when I use the system installed OpenSSL.

Solved by configuring OpenSSL to use the trusted certs from the directory where my system stores them and which is used by the system OpenSSL:
```common-lisp
(cl+ssl:ensure-initialized) ; needed to set the *ssl-global-context* 
(cl+ssl::ssl-ctx-set-verify-location cl+ssl::*ssl-global-context* "/etc/ssl/certs/")
```

**Pull request merged**